juniper/src/executor/look_ahead_lazy.rs

```diff
@@ -0,0 +1,1555 @@
+        self.input_list
+            .iter()
+            .map(|val| LookAheadValue::from_input_value(&val.item, &val.span, self.vars))
+    }
+}
+
+impl<'a, S> Default for LookAheadList<'a, S> {
+    fn default() -> Self {
+        Self {
+            input_list: &[],
+            vars: None,
+        }
+    }
+}
+
+impl<'a, S: ScalarValue> PartialEq for LookAheadList<'a, S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.iter().eq(other.iter())
+    }
+}
+
+/// A JSON-like object that can be used as an argument in the query execution.
+#[derive(Clone, Copy, Debug)]
+pub struct LookAheadObject<'a, S> {
+    input_object: &'a [(Spanning<String>, Spanning<InputValue<S>>)],
+    vars: Option<&'a Variables<S>>,
+}
+
+impl<'a, S: ScalarValue + 'a> LookAheadObject<'a, S> {
+    /// Returns an iterator over the object's entries.
+    pub fn iter(
+        &self,
+    ) -> impl Iterator<
+        Item = (
+            BorrowedSpanning<'a, &'a str>,
+            BorrowedSpanning<'a, LookAheadValue<'a, S>>,
+        ),
+    > + '_ {
+        self.input_object.iter().map(|(key, val)| {
+            (
+                Spanning {
+                    span: &key.span,
+                    item: key.item.as_str(),
+                },
+                LookAheadValue::from_input_value(&val.item, &val.span, self.vars),
+            )
+        })
+    }
+}
+
+impl<'a, S> Default for LookAheadObject<'a, S> {
+    fn default() -> Self {
+        Self {
+            input_object: &[],
+            vars: None,
+        }
+    }
+}
+
+impl<'a, S: ScalarValue> PartialEq for LookAheadObject<'a, S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.iter().eq(other.iter())
+    }
+}
+
+/// An argument passed into the query
+#[derive(Clone, Copy, Debug)]
+pub struct LookAheadArgument<'a, S> {
+    name: &'a Spanning<&'a str>,
+    input_value: &'a Spanning<InputValue<S>>,
+    vars: &'a Variables<S>,
+}
+
+impl<'a, S: ScalarValue> LookAheadArgument<'a, S> {
+    /// The argument's name
+    pub fn name(&self) -> &'a str {
+        self.name.item
+    }
+
+    /// The Span of the argument's name
+    pub fn name_span(&self) -> &'a Span {
+        &self.name.span
+    }
+
+    /// The argument's value
+    pub fn value(&self) -> LookAheadValue<'a, S> {
+        LookAheadValue::from_input_value(
+            &self.input_value.item,
+            &self.input_value.span,
+            Some(self.vars),
+        )
+        .item
+    }
+
+    /// The Span of the argument's value
+    pub fn value_span(&self) -> &'a Span {
+        &self.input_value.span
+    }
+}
+
+/// The arguments passed into a query.
+#[derive(Copy, Clone, Debug)]
+pub struct LookAheadArguments<'a, S> {
+    arguments: &'a Arguments<'a, S>,
+    vars: &'a Variables<S>,
+}
+
+impl<'a, S> LookAheadArguments<'a, S> {
+    /// Returns an iterator over the arguments.
+    pub fn iter(&self) -> impl Iterator<Item = LookAheadArgument<'a, S>> + '_ {
```

Comment:
It might be odd that `LookAheadChildren` implements `IntoIterator` but `LookAheadArguents` does not.

The reason is that the former owns its data, but the latter uses a slice with a `.map` operation, and we'd need a new type for the `type IntoIter`.

Reply:
@audunhalland there are some inconsistencies, though:
- `LookAheadArguments::iter()` returns `Iterator<Item = LookAheadArgument<'a, S>>` over owned values.
- `LookAheadChildren::iter()` returns `Iterator<Item = &LookAheadSelection<'a, S>>` over references.
- `<LookAheadChildren as IntoIterator>::into_iter()` returns `Iterator<Item = LookAheadSelection<'a, S>>` over owned values.

I suggest that `T::iter()` inherent method and `<&T as IntoIterator>::into_iter()` should always return `Iterator<Item = &T>` over references.

For owned values `Iterator<Item = T>` there should be either `<T as IntoIterator>::into_iter()` impl or  `T::into_iter()` inherent method.

Reply:
It's not so easy to return iterators over references when the iterated item is lazily computed. We're iterating over things that _wrap_ references. `LookAheadArguments` has `arguments: &Arguments` internally, but this isn't exposed directly. Instead each argument gets turned into a `LookAheadArgument` lazily as we iterate over them. If we wanted to iterate over `&LookAheadArgument` references, we'd have to preallocate a collection of those, which to me just seems wasteful without any particular gain.

Reply:
@audunhalland 

> It's not so easy to return iterators over references when the iterated item is lazily computed.

I do understand this concern. My comment is about naming only. If we cannot return iterators over references, then let's just name the methods `into_iter()` or impl the `IntoIterator` trait instead.

Reply:
Regarding `LookAheadChildren::iter(&self)` and that it yields borrowed `&LookAheadSelection` (this being inconsistent with some of the other `iter(&self)` iterators); this inconsistency is a reflection of how lazy_look_ahead is currently implemented. `LookAheadSelection` is not cheaply clonable because it owns a `Vec`. In Rust it's not considered bad to leak implementation details about how memory is managed internally. I of course agree that it makes it harder to refactor details later without breaking any public API, but this is just Rust.

Reply:
@audunhalland unfortunately we don't have exact this situation described in [guidelines](https://rust-lang.github.io/api-guidelines/naming.html#methods-on-collections-that-produce-iterators-follow-iter-iter_mut-into_iter-c-iter). Your vision that `iter`/`iter_mut` is about notating receiver rather than emitted values seems quite reasonable and logic.

However, it still feels quite odd to me to expect owned values as the result of `.iter()`. I don't know why. Usually, I expect some other naming like `.iter_values()` or just `.values()` notating exactly that some computation happens, and something other is produced. However, here it makes no sense, since the `LookAheadArguments` type itself means iteration over `LookAheadArgument`. There is no "other computed stuff".

Something like `.iter_owned()` is odd and not quite expected from the user-side too.

What if we impl `IntoIterator for &LookAheadArguments` with `type Item = LookAheadArgument`?

It should work relatively the same in method chains. As a bonus, it allows us `for arg in &args {}`.

Reply:
Suggestion for middle ground: `LookAheadValue`, `LookAheadList` and `LookAheadObject` all implement `Copy`, and there is no `iter(&self)` method at all on these types, only `IntoIterator` implementations for `Self` and `&Self`:

```
match look_ahead_value {
    LookAheadValue::List(list) => {
        list.into_iter().map(|value| { /* read values */ })
    }
    ...
}
```

The drawback is that this style would _likely_ be more surprising to the user. As we can see with the [clippy lint](https://github.com/rust-lang/rust-clippy/issues/1565) that suggests changing `.into_iter()` to `.iter()` if there is no actual ownership transfer going on, users are taught to reach for `.iter()` first, and a _GraphQL look ahead_ context is quite clearly a _borrow only_ context.

I'd think "hmm weird, why do I have to call .into_iter(), my mental model expects `.iter()` here". but I'd probably accept the situation and just move on with my life :)

Iteration methods can always be added at a later stage if users turn out to find this confusing. Naming suggestion in that case:
* `fn LookAheadList::elements(&self)`
* `fn LookAheadObject::entries(&self)` (although this might be confusing as I would think of Rust's Map entry API.)